File: PrebidMobile/Rendering/EventHandlers/PrebidMobileMoPubAdapters/Sources/PrebidMoPubRewardedVideoAdapter.m
//
//  PBMMoPubRewardedVideoAdapter.m
//  OpenXInternalTestApp
//
//  Copyright © 2020 OpenX. All rights reserved.
//
#import <MoPub/MoPub.h>

#import <PrebidMobileRendering/PBMInterstitialController.h>

#import "PrebidMoPubRewardedVideoAdapter.h"

@interface PrebidMoPubRewardedVideoAdapter () <PBMInterstitialControllerLoadingDelegate, PBMInterstitialControllerInteractionDelegate>

@property (nonatomic, strong) PBMInterstitialController* interstitialController;
@property (nonatomic, weak) UIViewController *rootViewController;
@property (nonatomic, copy) NSString *configId;
@property (nonatomic, assign) BOOL adAvailable;
@end

@implementation PrebidMoPubRewardedVideoAdapter

- (void)requestAdWithAdapterInfo:(NSDictionary *)info adMarkup:(NSString *)adMarkup {
    if (self.localExtras.count == 0) {
        NSError *error = [NSError errorWithDomain:PBMErrorDomain
                                             code:PBMErrorCodeGeneral
                                         userInfo:@{NSLocalizedDescriptionKey: NSLocalizedString(@"The local extras is empty", nil)}];
        MPLogAdEvent([MPLogEvent adLoadFailedForAdapter:NSStringFromClass(self.class) error:error], @"");
        [self.delegate fullscreenAdAdapter:self didFailToLoadAdWithError:error];
        return;
    }
    
    PBMBid *bid = self.localExtras[PBMMoPubAdUnitBidKey];
    if (!bid) {
        NSError *error = [NSError errorWithDomain:PBMErrorDomain
                                             code:PBMErrorCodeGeneral
                                         userInfo:@{NSLocalizedDescriptionKey: NSLocalizedString(@"The Bid object is absent in the extras", nil)}];
        MPLogAdEvent([MPLogEvent adLoadFailedForAdapter:NSStringFromClass(self.class) error:error], [self getAdUnitId]);
        [self.delegate fullscreenAdAdapter:self didFailToLoadAdWithError:error];
        return;
    }
    
    self.configId = self.localExtras[PBMMoPubConfigIdKey];
    if (!self.configId) {
        NSError *error = [NSError errorWithDomain:PBMErrorDomain
                                             code:PBMErrorCodeGeneral
                                         userInfo:@{NSLocalizedDescriptionKey: NSLocalizedString(@"The Config ID absent in the extras", nil)}];
        MPLogAdEvent([MPLogEvent adLoadFailedForAdapter:NSStringFromClass(self.class) error:error], [self getAdUnitId]);
        [self.delegate fullscreenAdAdapter:self didFailToLoadAdWithError:error];
        return;
    }
    
    self.interstitialController = [[PBMInterstitialController alloc] initWithBid:bid configId:self.configId];
    self.interstitialController.loadingDelegate = self;
    self.interstitialController.interactionDelegate = self;
    self.interstitialController.adFormat = PBMAdFormatVideo;
    self.interstitialController.isOptIn = YES;
    
    [self.interstitialController loadAd];
    MPLogAdEvent([MPLogEvent adLoadAttemptForAdapter:NSStringFromClass(self.class) dspCreativeId:nil dspName:nil], [self getAdUnitId]);
}

-(BOOL)isRewardExpected {
    return YES;
}

-(BOOL)hasAdAvailable {
    return self.adAvailable;
}

- (void)presentAdFromViewController:(UIViewController *)viewController {
    MPLogAdEvent([MPLogEvent adShowAttemptForAdapter:NSStringFromClass(self.class)], [self getAdUnitId]);
    
    if ([self hasAdAvailable]) {
        self.rootViewController = viewController;
        [self.interstitialController show];
    } else {
        NSError *error = [NSError errorWithDomain:PBMErrorDomain
                                             code:PBMErrorCodeGeneral
                                         userInfo:@{NSLocalizedDescriptionKey: NSLocalizedString(@"The ad hasn’t been loaded", nil)}];
        MPLogAdEvent([MPLogEvent adShowFailedForAdapter:NSStringFromClass(self.class) error:error], [self getAdUnitId]);
        [self.delegate fullscreenAdAdapter:self didFailToLoadAdWithError:error];
    }
}

- (NSString *) getAdUnitId {
    return self.configId ?: @"";
}

#pragma mark - PBMInterstitialControllerDelegate

- (void)interstitialController:(nonnull PBMInterstitialController *)interstitialController didFailWithError:(nonnull NSError *)error {
    MPLogAdEvent([MPLogEvent adLoadFailedForAdapter:NSStringFromClass(self.class) error:error], [self getAdUnitId]);
    self.adAvailable = NO;
    [self.delegate fullscreenAdAdapter:self didFailToLoadAdWithError:error];
}

- (void)interstitialControllerDidDisplay:(PBMInterstitialController *)interstitialController {
    NSString *adUnitId = [self getAdUnitId];
    NSString *classString = NSStringFromClass(self.class);
    MPLogAdEvent(MPLogEvent.adShowSuccess, adUnitId);
    MPLogAdEvent([MPLogEvent adWillAppearForAdapter:classString], adUnitId);
    MPLogAdEvent([MPLogEvent adShowSuccessForAdapter:classString], adUnitId);
    MPLogAdEvent([MPLogEvent adDidAppearForAdapter:classString], adUnitId);
    
    [self.delegate fullscreenAdAdapterAdWillAppear:self];
    [self.delegate fullscreenAdAdapterAdDidAppear:self];
}

- (void)interstitialControllerDidComplete:(PBMInterstitialController *)interstitialController {
    self.adAvailable = NO;
    
    // Get rid of the interstitial view controller when done with it so we don't hold on longer than needed
    self.interstitialController = nil;
    MPLogInfo(@"Interstitial did complete");
    MPReward *reward = [[MPReward alloc] initWithCurrencyAmount:@(1)];
    [self.delegate fullscreenAdAdapter:self willRewardUser:reward];
}

- (void)interstitialControllerDidClickAd:(nonnull PBMInterstitialController *)interstitialController {
    MPLogAdEvent([MPLogEvent adTappedForAdapter:NSStringFromClass(self.class)], [self getAdUnitId]);
    [self.delegate fullscreenAdAdapterDidReceiveTap:self];
}

- (void)interstitialControllerDidCloseAd:(nonnull PBMInterstitialController *)interstitialController {
    self.adAvailable = NO;
    NSString *adUnitId = [self getAdUnitId];
    MPLogAdEvent([MPLogEvent adWillDisappearForAdapter:NSStringFromClass(self.class)], adUnitId);
    MPLogAdEvent([MPLogEvent adDidDisappearForAdapter:NSStringFromClass(self.class)], adUnitId);

    [self.delegate fullscreenAdAdapterAdWillDisappear:self];
    [self.delegate fullscreenAdAdapterAdDidDisappear:self];
}

- (void)interstitialControllerDidLeaveApp:(nonnull PBMInterstitialController *)interstitialController {
    MPLogAdEvent([MPLogEvent adWillLeaveApplicationForAdapter:NSStringFromClass(self.class)], [self getAdUnitId]);
    [self.delegate fullscreenAdAdapterWillLeaveApplication:self];
}

- (void)interstitialControllerDidLoadAd:(nonnull PBMInterstitialController *)interstitialController {
    MPLogAdEvent([MPLogEvent adLoadSuccessForAdapter:NSStringFromClass(self.class)], [self getAdUnitId]);
    self.adAvailable = YES;
    [self.delegate fullscreenAdAdapterDidLoadAd:self];
}

- (void)trackImpressionForInterstitialController:(nonnull PBMInterstitialController *)interstitialController {
    //Impressions will be tracked automatically
    //unless enableAutomaticImpressionAndClickTracking = NO
    //In this case you have to override the didDisplayAd method
    //and manually call inlineAdAdapterDidTrackImpression
    //in this method to ensure correct metrics
}

- (nonnull UIViewController *)viewControllerForModalPresentationFrom:(nonnull PBMInterstitialController *)interstitialController {
    return self.rootViewController;
}


@end

File: EventHandlers/PrebidMobileMoPubAdapters/Sources/PrebidMoPubRewardedVideoAdapter.swift
/*   Copyright 2018-2021 Prebid.org, Inc.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

import Foundation

import MoPubSDK

import PrebidMobile

@objc(PrebidMoPubRewardedVideoAdapter)
public class PrebidMoPubRewardedVideoAdapter :
    MPFullscreenAdAdapter,
    InterstitialControllerLoadingDelegate,
    InterstitialControllerInteractionDelegate {
    
    // MARK: - Private Properties
    
    var interstitialController: InterstitialController?
    weak var rootViewController: UIViewController?
    var configID: String?
    var adAvailable = false
    
    // MARK: - MPFullscreenAdAdapter
    
    override public var hasAdAvailable: Bool {
        get { adAvailable }
        set { adAvailable = newValue }
    }
    
    public override func requestAd(withAdapterInfo info: [AnyHashable : Any], adMarkup: String?) {
        guard !localExtras.isEmpty else {
            let error = MoPubAdaptersError.emptyLocalExtras
            MPLogging.logEvent(MPLogEvent.adLoadFailed(forAdapter: Self.className(), error: error), source: adUnitId, from: nil)
            delegate?.fullscreenAdAdapter(self, didFailToLoadAdWithError: error)
            return
        }
        
        guard let bid = localExtras[PBMMoPubAdUnitBidKey] as? Bid else {
            let error = MoPubAdaptersError.noBidInLocalExtras
            MPLogging.logEvent(MPLogEvent.adLoadFailed(forAdapter: Self.className(), error: error), source: adUnitId, from: nil)
            delegate?.fullscreenAdAdapter(self, didFailToLoadAdWithError: error)
            return
        }
        
        guard let configID = localExtras[PBMMoPubConfigIdKey] as? String else {
            let error = MoPubAdaptersError.noConfigIDInLocalExtras
            MPLogging.logEvent(MPLogEvent.adLoadFailed(forAdapter: Self.className(), error: error), source: adUnitId, from: nil)
            delegate?.fullscreenAdAdapter(self, didFailToLoadAdWithError: error)
            return
        }
        
        interstitialController = InterstitialController(bid: bid, configId: configID)
        interstitialController?.loadingDelegate = self
        interstitialController?.interactionDelegate = self
        interstitialController?.adFormat = .video
        interstitialController?.isOptIn = true
        
        interstitialController?.loadAd()
        
        MPLogging.logEvent(MPLogEvent.adLoadAttempt(forAdapter: Self.className(), dspCreativeId: nil, dspName: nil), source: adUnitId, from: nil)
    }
    
    public override func presentAd(from viewController: UIViewController) {
        MPLogging.logEvent(MPLogEvent.adShowAttempt(forAdapter: Self.className()), source: adUnitId, from: nil)

        if hasAdAvailable {
            rootViewController = viewController
            interstitialController?.show()
        } else {
            let error = MoPubAdaptersError.noAd
            MPLogging.logEvent(MPLogEvent.adLoadFailed(forAdapter: Self.className(), error: error), source: adUnitId, from: nil)
            delegate?.fullscreenAdAdapter(self, didFailToLoadAdWithError: error)
        }
    }
    
    public override var isRewardExpected: Bool {
        true
    }
        
    // MARK: - InterstitialControllerLoadingDelegate
    
    public func interstitialControllerDidLoadAd(_ interstitialController: InterstitialController) {
        adAvailable = true
        
        MPLogging.logEvent(MPLogEvent.adLoadSuccess(forAdapter: Self.className()), source: adUnitId, from: nil)
        delegate?.fullscreenAdAdapterDidLoadAd(self)
    }
    
    public func interstitialController(_ interstitialController: InterstitialController, didFailWithError error: Error) {
        adAvailable = false
        
        MPLogging.logEvent(MPLogEvent.adLoadFailed(forAdapter: Self.className(), error: error), source: adUnitId, from: nil)
        delegate?.fullscreenAdAdapter(self, didFailToLoadAdWithError: error)
    }
    
    // MARK: - InterstitialControllerInteractionDelegate
    
    public func trackImpression(for interstitialController: InterstitialController) {
        //Impressions will be tracked automatically
        //unless enableAutomaticImpressionAndClickTracking = NO
        //In this case you have to override the didDisplayAd method
        //and manually call inlineAdAdapterDidTrackImpression
        //in this method to ensure correct metrics
    }
    
    public func interstitialControllerDidClickAd(_ interstitialController: InterstitialController) {
        MPLogging.logEvent(MPLogEvent.adTapped(forAdapter: Self.className()), source: adUnitId, from: nil)
        delegate?.fullscreenAdAdapterDidReceiveTap(self)
    }
    
    public func interstitialControllerDidCloseAd(_ interstitialController: InterstitialController) {
        adAvailable = false
        
        MPLogging.logEvent(MPLogEvent.adWillDisappear(forAdapter: Self.className()), source: adUnitId, from: nil)
        MPLogging.logEvent(MPLogEvent.adDidDisappear(forAdapter: Self.className()), source: adUnitId, from: nil)

        delegate?.fullscreenAdAdapterAdWillDisappear(self)
        delegate?.fullscreenAdAdapterAdDidDisappear(self)
    }
    
    public func interstitialControllerDidLeaveApp(_ interstitialController: InterstitialController) {
        MPLogging.logEvent(MPLogEvent.adWillLeaveApplication(forAdapter: Self.className()), source: adUnitId, from: nil)
        delegate?.fullscreenAdAdapterWillLeaveApplication(self)
    }
    
    public func viewControllerForModalPresentation(from interstitialController: InterstitialController) -> UIViewController? {
        rootViewController
    }
    
    public func interstitialControllerDidDisplay(_ interstitialController: InterstitialController) {
        let className = Self.className()
        
        MPLogging.logEvent(MPLogEvent.adWillAppear(forAdapter: className), source: adUnitId, from: nil)
        MPLogging.logEvent(MPLogEvent.adDidAppear(forAdapter: className), source: adUnitId, from: nil)
        MPLogging.logEvent(MPLogEvent.adShowSuccess(forAdapter: className), source: adUnitId, from: nil)

        delegate?.fullscreenAdAdapterAdWillAppear(self)
        delegate?.fullscreenAdAdapterAdDidAppear(self)
    }
    
    public func interstitialControllerDidComplete(_ interstitialController: InterstitialController) {
        adAvailable = false
        self.interstitialController = nil
        
        let reward = MPReward()
        delegate?.fullscreenAdAdapter(self, willRewardUser: reward)
    }
}
